src/components/vm/filesystems/vmFilesystemsCard.jsx

```diff
@@ -0,0 +1,254 @@
+import { vmId } from "../../../helpers.js";
+
+import "./vmFilesystemsCard.scss";
+
+const _ = cockpit.gettext;
+
+export const VmFilesystemsCard = ({ vmName, filesystems }) => {
+    const columnTitles = [_("Source path"), _("Mount tag")];
+
+    const rows = filesystems.map(filesystem => {
+        const sourceKey = Object.keys(filesystem.source).find(key => filesystem.source[key]);
+        const filesystemSource = sourceKey ? filesystem.source[sourceKey] : undefined;
+        const filesystemTarget = filesystem.target.dir;
+
+        const columns = [
+            { title: filesystemSource },
+            { title: filesystemTarget },
+        ];
+        return {
+            columns,
+            props: {
+                key: cockpit.format("$0-filesystem-$1-$2", vmId(vmName), filesystemSource, filesystemTarget),
+                'data-row-id': cockpit.format("$0-filesystem-$1-$2", vmId(vmName), filesystemSource, filesystemTarget),
+            }
+        };
+    });
+
+    return (
+        <ListingTable variant='compact'
+                      gridBreakPoint='grid-xl'
+                      emptyCaption={_("No directories shared between the host and this VM")}
+                      columns={columnTitles}
+                      rows={rows} />
+    );
+};
+
+export const VmFilesystemActions = ({ connectionName, dispatch, objPath, vmName, vmState, memory, memoryBacking }) => {
+    const [additionalOptionsExpanded, setAdditionalOptionsExpanded] = useState(false);
+    const [dialogError, setDialogError] = useState();
+    const [isOpen, setIsOpen] = useState(false);
+    const [memoryBackingType, setMemoryBackingType] = useState("file");
+    const [mountTag, setMountTag] = useState("");
+    const [source, setSource] = useState("");
+    const [validationFailed, setValidationFailed] = useState({});
+    const [xattr, setXattr] = useState(false);
+    const idPrefix = `${vmId(vmName)}-filesystems`;
+
+    const onAddClicked = () => {
+        const validationFailed = {};
+
+        if (!mountTag)
+            validationFailed.mountTag = _("Mount tag must not be empty");
+        if (!source)
+            validationFailed.source = _("Source must not be empty");
+
+        setValidationFailed(validationFailed);
+
+        if (Object.getOwnPropertyNames(validationFailed).length == 0) {
+            setMemoryBacking({
+                connectionName, objPath,
+                type: memoryBackingType,
+                memory
+            })
+                    .then(() => {
+                        createFilesystem({
+                            connectionName, objPath,
+                            source, target: mountTag,
+                            xattr,
+                        });
+                    })
+                    .then(
+                        () => {
+                            setAdditionalOptionsExpanded(false);
+                            setDialogError();
+                            setIsOpen(false);
```

Comment:
This seems to reset the dialog back to the defaults. But why does that happen after creating the fs? Shouldn't that be done when clicking the "Add shared directory" button rather? Also, why is this reset necessary in the first place? Wouldn't it be cleaner to completely destroy the component after closing the dialog, and reinstantiate a new one when clicking the button? Then the defaults don't need to be repeated (both here and in `useState()` calls).

Reply:
Moved the modal to a separate component and so I just recreate the component when closed.